### PR TITLE
Significantly tighten memory usage in build_from

### PR DESCRIPTION
**Pull request by cmc333333**
This improves memory usage (allowing reg Z to compile on machines with < 1G memory) through a few optimizations.

1. FrozenNodes now have a proper digest hash, which is calculated at their creation time. Equality checks look only at the fields and the hashes, reducing the need to generate a gigantic `repr` string.
1. FrozenNode creation should only occur once. Because they are immutable, we don't need duplicate versions of the same FrozenNode. This means that the majority of the tree is shared between versions which cuts memory usage by about half (for Z)
1. The build_from script includes a few more `del` statements, to indicate memory can be reclaimed.

I also synchronized the build_from script with CFPB's re-order. That'll make merging upstream much easier.

### CODE REVIEW

**Comment by cmc333333** [comment]
Bump @khandelwal

**Comment by khandelwal** [comment]
Sorry - I'll take a look at this later today.

File: regparser/tree/struct.py [hunk 1]
```diff
@@ -175,27 +181,53 @@ def node_type(self):
...
+        """As the hash property is already distinctive, re-use it"""
+        return hash(self.hash)
+
+    def __eq__(self, other):
+        """We define equality as having the same fields except for children.
+        Instead of recursively inspecting them, we compare only their hash
+        (this is a Merkle tree)"""
+        return (other.__class__ == self.__class__
+                and self.hash == other.hash
+                and self.text == other.text
```
Comment (khandelwal) [inline regparser/tree/struct.py]: The hash already includes all these other things, such as text, title. Why do we need to compare those here? Isn't comparing the hash enough? I'm obviously missing something...

Comment (cmc333333) [inline regparser/tree/struct.py]: Added a comment to explain; there's a very small chance of a hash collision -- that two different `FrozenNodes` would hash to the same entity. This limits the damage that might do by also comparing fields.

**Comment by cmc333333** [comment]
Bump

**Pull request merged**